### ISSUE TITLE
Issues with JoinHandle's poll implementation not waking after finishing, and unsoundness in file.rs.

### ISSUE THREAD

**Issue by CoolOppo** [async-rs/async-std#683]
I was reading through [file.rs](https://github.com/async-rs/async-std/blob/46cafffc31711d5df1b267dc38d191463ac10bda/src/fs/file.rs) trying to figure out how exactly this stuff works on the inside. I think I found a couple of problems with reading files and `JoinHandle`, **not entirely sure**. I may be missing a detail or two that explain this all away, but it seems possible enough that I found some issues here worth bringing up that I'm willing to look stupid if I'm completely wrong. I'm going to ping @stjepang and @yoshuawuyts as it seems you two are the ones who actually wrote this.

First off is something suspicious on [line 673](https://github.com/async-rs/async-std/blob/46cafffc31711d5df1b267dc38d191463ac10bda/src/fs/file.rs#L673). Seems like we shouldn't run the following code on `is_empty()`. Probably should be an immediate `return Poll::Ready(Ok(0));` in that scenario.

On [line 703](https://github.com/async-rs/async-std/blob/46cafffc31711d5df1b267dc38d191463ac10bda/src/fs/file.rs#L703), a blocking task is spawned and the `JoinHandle` is immediately dropped. It isn't ever properly polled. I am assuming we at least don't poll *immediately* because it will never have been completed anyways, so it's not worth the cost of pinning it and returning its `Poll` result (because it will always just be `Poll::Pending` unless a miracle occurs.)

Without the ability for the `JoinHandle` to actually call `wake` when its `Task` is done, we have the current situation. Basically, after we start reading to the cache with `spawn_blocking(...)`, we're done running `poll_read(...)`, so we drop the `LockGuard`, which will unlock the `Mutex` and wake up everyone who was `register`ed as a "listener" on the file lock. In the case that there isn't anything else going on for our executor to run, we're then immediately going to be woken back up, the lock is going to be attained again, and we're going to check if there is anything to read from the cache. If there isn't, we just keep dropping, unlocking, waking, locking, and checking repeatedly until there is.

What it seems *should* be happening is that when the `JoinHandle` from `spawn_blocking(...)` is `poll`ed, it would, [like the docs say](https://docs.rs/async-std/1.4.0/async_std/future/trait.Future.html#tymethod.poll), call `Waker::wake` on the provided `Context`'s `Waker` when it has completed. The problem is that `JoinHandle` doesn't do this and doesn't store the waker anywhere like a `Future` is supposed to. [It just returns `Poll::Pending` and that's it](https://github.com/async-rs/async-std/blob/87de4e15982bbec890aa909600da5997c4f6edc7/src/task/join_handle.rs#L46). Then we could get rid of the waking that occurs when dropping the LockGuard and the blocking task will just do the waking.

Another possibility would be, I think, using actual platform-specific events for IO, although the issues with `JoinHandle` would still exist for users of `spawn_blocking(...)`

There also seem to be a couple other holes in the file read logic. This may be worthy for another issue. If you think so, I'll move the rest of what's here over.

Since we're setting the cache's length and reading into it, and the cache is the same for all operations on the file, we could be changing this while another read is happening or even writing to the cache two times at once, interleaving the data. After all, we only have mutable access to the cache because the `LockGuard` has [`unsafe` implementations of `Deref`](https://github.com/async-rs/async-std/blob/master/src/fs/file.rs#L550-L562). With the current way the logic associated with locking works, this would be extremely rare because basically another poll_read would have to happen before we got to the point of registering ourselves before them. The chances of that are very slim but it's still possible. It would be a much larger issue though if you go the route of having the `JoinHandle` call `wake` when it's done, because the order of the tasks being woken wouldn't just be FRFW (First Registered First Woke ;) ) any more.

------

Anyways, thanks for your work on the project so far. It's a nice codebase I have been able to traverse and understand very easily! It's far easier to read this than *tokio* and [*asio*](https://think-async.com/Asio/). If not for the structure of this project I wouldn't have been able to find these things in the first place. I'm a beginner with async and have only been using rust for about a year, so the fact that I'm able to delve into this codebase should speak volumes about its quality.

**Comment by skade** [comment]
Hi @CoolOppo, thanks for this _thorough_ report and spending time reading and noting down your findings!

**Comment by CoolOppo** [comment]
No problem; learned a lot about how things actually work below the surface by delving in here. If it weren't for the quality of the codebase, I would never have even had the patience to try and examine what was going on in the first place, let alone fester up enough confidence about how it works to think "maybe this is actually worth asking about".

**Comment by stjepang** [comment]
> First off is something suspicious on [line 673](https://github.com/async-rs/async-std/blob/46cafffc31711d5df1b267dc38d191463ac10bda/src/fs/file.rs#L673). Seems like we shouldn't run the following code on `is_empty()`. Probably should be an immediate `return Poll::Ready(Ok(0));` in that scenario.

That's exactly what will happen since `n` will be 0. :) If the cache is empty, we can  "successfully" read 0 bytes from it.

> On [line 703](https://github.com/async-rs/async-std/blob/46cafffc31711d5df1b267dc38d191463ac10bda/src/fs/file.rs#L703), a blocking task is spawned and the `JoinHandle` is immediately dropped. It isn't ever properly polled. I am assuming we at least don't poll _immediately_ because it will never have been completed anyways, so it's not worth the cost of pinning it and returning its `Poll` result (because it will always just be `Poll::Pending` unless a miracle occurs.)

Note that the `LockGuard` (variable `self`) is moved into the `spawn_blocking` closure. Before calling `spawn_blocking`, we register the current tasks's interest in the lock (it's like polling the lock) at line 700. Once the blocking closure completes, it will drop the `LockGuard` and wake the original task.

> Since we're setting the cache's length and reading into it, and the cache is the same for all operations on the file, we could be changing this while another read is happening or even writing to the cache two times at once, interleaving the data.

That is true. Unfortunately, I don't see a way to prevent the problem. I believe this problem is inherent to the way `AsyncRead` and `AsyncWrite` work. As you say, it is unlikely to happen. If the problem does occur, it is not unreasonable to blame the user for it, I think. :)

**Comment by CoolOppo** [comment]
Hey, thanks for the prompt answers! 

> Note that the `LockGuard` (variable `self`) is moved into the `spawn_blocking` closure. [...] Once the blocking closure completes, it will drop the `LockGuard` and wake the original task.

I can't believe I missed this! Everything comes together now knowing this! 😁

I still feel the `JoinHandle` future doesn't behave the way it's "supposed" to and call the Waker when it's completed. It would make it easier to understand and more useful to library consumers if it did. I hope it's something that'll be ironed out in the future. Still, no clue how far that should be up your priority list (especially considering `spawn_blocking` is experimental in itself I believe?) so there are no serious complaints from me in this regard. Really. It's a great framework you folks threw together here.

> That is true. Unfortunately, I don't see a way to prevent the problem. I believe this problem is inherent to the way `AsyncRead` and `AsyncWrite` work. As you say, it is unlikely to happen. If the problem does occur, it is not unreasonable to blame the user for it, I think. :)

At least I got something right! Lol. I do feel as though there should be some kind of warning in the docs about performing ops that might result in something unexpected. An issue in reading files, even super rare, could turn into a seriously frustrating day for some poor sap consuming this API if they aren't aware of this.
